File: references/java-sdk/src/main/java/blue/language/sdk/AgencyBuilder.java
package blue.language.sdk;

import blue.language.sdk.internal.StepsBuilder;
import blue.language.types.myos.WorkerAgencyPermissionGranted;
import blue.language.types.myos.WorkerAgencyPermissionRejected;
import blue.language.types.myos.WorkerAgencyPermissionRevoked;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.Consumer;

public final class AgencyBuilder<P extends DocBuilder<P>> {

    private final P parent;
    private final String name;
    private String onBehalfOfChannel;
    private final List<Class<?>> allowedTypes = new ArrayList<Class<?>>();
    private final List<String> allowedOperations = new ArrayList<String>();
    private String statusPath;

    private AccessConfig.PermissionTiming permissionTiming = AccessConfig.PermissionTiming.ON_INIT;
    private Class<?> permissionTriggerEventClass;
    private String permissionTriggerDocPath;

    AgencyBuilder(P parent, String name) {
        if (parent == null) {
            throw new IllegalArgumentException("parent is required");
        }
        if (name == null || name.trim().isEmpty()) {
            throw new IllegalArgumentException("agency name is required");
        }
        this.parent = parent;
        this.name = name.trim();
    }

    public AgencyBuilder<P> onBehalfOf(String channelKey) {
        this.onBehalfOfChannel = channelKey;
        return this;
    }

    public AgencyBuilder<P> allowedTypes(Class<?>... allowedTypes) {
        if (allowedTypes != null) {
            this.allowedTypes.addAll(Arrays.asList(allowedTypes));
        }
        return this;
    }

    public AgencyBuilder<P> allowedOperations(String... allowedOperations) {
        if (allowedOperations != null) {
            this.allowedOperations.addAll(Arrays.asList(allowedOperations));
        }
        return this;
    }

    public AgencyBuilder<P> statusPath(String statusPath) {
        this.statusPath = statusPath;
        return this;
    }

    public AgencyBuilder<P> requestPermissionOnInit() {
        this.permissionTiming = AccessConfig.PermissionTiming.ON_INIT;
        this.permissionTriggerEventClass = null;
        this.permissionTriggerDocPath = null;
        return this;
    }

    public AgencyBuilder<P> requestPermissionOnEvent(Class<?> eventTypeClass) {
        if (eventTypeClass == null) {
            throw new IllegalArgumentException("eventTypeClass is required");
        }
        this.permissionTiming = AccessConfig.PermissionTiming.ON_EVENT;
        this.permissionTriggerEventClass = eventTypeClass;
        this.permissionTriggerDocPath = null;
        return this;
    }

    public AgencyBuilder<P> requestPermissionOnDocChange(String path) {
        if (path == null || path.trim().isEmpty()) {
            throw new IllegalArgumentException("path is required");
        }
        this.permissionTiming = AccessConfig.PermissionTiming.ON_DOC_CHANGE;
        this.permissionTriggerDocPath = path.trim();
        this.permissionTriggerEventClass = null;
        return this;
    }

    public AgencyBuilder<P> requestPermissionManually() {
        this.permissionTiming = AccessConfig.PermissionTiming.MANUAL;
        this.permissionTriggerEventClass = null;
        this.permissionTriggerDocPath = null;
        return this;
    }

    public P done() {
        validate();

        String token = DocBuilder.tokenOf(name, "AGENCY");
        String requestId = "REQ_AGENCY_" + token;

        AgencyConfig config = new AgencyConfig(
                name,
                token,
                onBehalfOfChannel,
                requestId,
                allowedTypes,
                allowedOperations,
                statusPath,
                permissionTiming,
                permissionTriggerEventClass,
                permissionTriggerDocPath);
        parent.registerAgencyConfig(config);

        if (config.statusPath() != null) {
            parent.field(config.statusPath(), "pending");
        }

        if (config.permissionTiming() != AccessConfig.PermissionTiming.MANUAL) {
            String workflowKey = "agency" + token + "RequestPermission";
            Consumer<StepsBuilder> permissionStep = steps -> steps.myOs().grantWorkerAgencyPermission(
                    config.onBehalfOfChannel(),
                    config.requestId(),
                    config.permissionNode());
            switch (config.permissionTiming()) {
                case ON_INIT:
                    parent.onInit(workflowKey, permissionStep);
                    break;
                case ON_EVENT:
                    parent.onEvent(workflowKey, config.permissionTriggerEventClass(), permissionStep);
                    break;
                case ON_DOC_CHANGE:
                    parent.onDocChange(workflowKey, config.permissionTriggerDocPath(), permissionStep);
                    break;
                default:
                    break;
            }
        }

        parent.onMyOsResponse("agency" + token + "Granted",
                WorkerAgencyPermissionGranted.class,
                config.requestId(),
                steps -> {
                    if (config.statusPath() != null) {
                        steps.replaceValue("MarkAgencyGranted", config.statusPath(), "granted");
                    }
                });

        parent.onMyOsResponse("agency" + token + "Rejected",
                WorkerAgencyPermissionRejected.class,
                config.requestId(),
                steps -> {
                    if (config.statusPath() != null) {
                        steps.replaceValue("MarkAgencyRejected", config.statusPath(), "rejected");
                    }
                });

        parent.onMyOsResponse("agency" + token + "Revoked",
                WorkerAgencyPermissionRevoked.class,
                config.requestId(),
                steps -> {
                    if (config.statusPath() != null) {
                        steps.replaceValue("MarkAgencyRevoked", config.statusPath(), "revoked");
                    }
                });

        return parent;
    }

    private void validate() {
        if (onBehalfOfChannel == null || onBehalfOfChannel.trim().isEmpty()) {
            throw new IllegalArgumentException("agency('" + name + "'): onBehalfOf is required");
        }
    }
}

File: libs/dsl-sdk/src/lib/interactions/agency-builder.ts
import { toTypeAlias, type TypeLike } from '../core/type-alias.js';
import type {
  AgencyConfig,
  AgencyConfigRegistrationHost,
  InteractionPermissionTiming,
} from './types.js';

function requireText(value: string | undefined, field: string): string {
  const normalized = value?.trim();
  if (!normalized) {
    throw new Error(`${field} is required`);
  }
  return normalized;
}

function token(name: string): string {
  const normalized = name
    .replace(/[^A-Za-z0-9]+/gu, '_')
    .replace(/^_+|_+$/gu, '')
    .toUpperCase();
  return normalized.length > 0 ? normalized : 'AGENCY';
}

type StepsCustomizer = (steps: {
  myOs: (adminChannelKey?: string) => {
    grantWorkerAgencyPermission: (
      onBehalfOf: string,
      requestId: string,
      workerAgencyPermissions: Record<string, unknown>,
      targetSessionId?: string,
    ) => unknown;
  };
  replaceValue: (name: string, path: string, value: unknown) => unknown;
}) => void;

type DocBuilderLike = {
  field(path: string, value: unknown): unknown;
  onInit(workflowKey: string, customizer: StepsCustomizer): unknown;
  onEvent(
    workflowKey: string,
    eventType: string,
    customizer: StepsCustomizer,
  ): unknown;
  onDocChange(
    workflowKey: string,
    path: string,
    customizer: StepsCustomizer,
  ): unknown;
  onMyOsResponse(
    workflowKey: string,
    responseType: string,
    requestId: string,
    customizer: StepsCustomizer,
  ): unknown;
};

export class AgencyBuilder<P> {
  private onBehalfOfValue: string | undefined;
  private requestIdValue: string | undefined;
  private targetSessionIdValue: string | undefined;
  private readonly allowedTypesValue: string[] = [];
  private readonly allowedOperationsValue: string[] = [];
  private statusPathValue: string | undefined;
  private permissionTimingValue: InteractionPermissionTiming = 'onInit';
  private permissionTriggerEventTypeValue: string | undefined;
  private permissionTriggerDocPathValue: string | undefined;

  constructor(
    private readonly parent: AgencyConfigRegistrationHost<P>,
    private readonly name: string,
  ) {}

  onBehalfOf(channelKey: string): this {
    this.onBehalfOfValue = requireText(channelKey, 'onBehalfOf');
    return this;
  }

  permissionFrom(channelKey: string): this {
    this.onBehalfOf(channelKey);
    return this;
  }

  requestId(requestId: string): this {
    this.requestIdValue = requireText(requestId, 'requestId');
    return this;
  }

  targetSessionId(targetSessionId: string): this {
    this.targetSessionIdValue = requireText(targetSessionId, 'targetSessionId');
    return this;
  }

  allowedTypes(...types: TypeLike[]): this {
    this.allowedTypesValue.length = 0;
    this.allowedTypesValue.push(...types.map((value) => toTypeAlias(value)));
    return this;
  }

  allowedOperations(...operations: string[]): this {
    this.allowedOperationsValue.length = 0;
    this.allowedOperationsValue.push(
      ...operations
        .map((operation) => operation?.trim())
        .filter((operation): operation is string => Boolean(operation)),
    );
    return this;
  }

  statusPath(statusPath: string): this {
    this.statusPathValue = requireText(statusPath, 'statusPath');
    return this;
  }

  requestPermissionOnInit(): this {
    this.permissionTimingValue = 'onInit';
    this.permissionTriggerEventTypeValue = undefined;
    this.permissionTriggerDocPathValue = undefined;
    return this;
  }

  requestPermissionOnEvent(eventType: TypeLike): this {
    this.permissionTimingValue = 'onEvent';
    this.permissionTriggerEventTypeValue = toTypeAlias(eventType);
    this.permissionTriggerDocPathValue = undefined;
    return this;
  }

  requestPermissionOnDocChange(path: string): this {
    this.permissionTimingValue = 'onDocChange';
    this.permissionTriggerDocPathValue = requireText(path, 'permission path');
    this.permissionTriggerEventTypeValue = undefined;
    return this;
  }

  requestPermissionManually(): this {
    this.permissionTimingValue = 'manual';
    this.permissionTriggerDocPathValue = undefined;
    this.permissionTriggerEventTypeValue = undefined;
    return this;
  }

  done(): P {
    const baseToken = token(this.name);
    const config: AgencyConfig = {
      name: this.name,
      token: baseToken,
      onBehalfOf: requireText(this.onBehalfOfValue, 'onBehalfOf'),
      requestId: this.requestIdValue ?? `REQ_AGENCY_${baseToken}`,
      allowedTypes: [...new Set(this.allowedTypesValue)],
      allowedOperations: [...new Set(this.allowedOperationsValue)],
      ...(this.statusPathValue ? { statusPath: this.statusPathValue } : {}),
      permissionTiming: this.permissionTimingValue,
      ...(this.permissionTriggerEventTypeValue
        ? { permissionTriggerEventType: this.permissionTriggerEventTypeValue }
        : {}),
      ...(this.permissionTriggerDocPathValue
        ? { permissionTriggerDocPath: this.permissionTriggerDocPathValue }
        : {}),
      permissionFrom: this.onBehalfOfValue,
      ...(this.targetSessionIdValue
        ? { targetSessionId: this.targetSessionIdValue }
        : {}),
    };
    const registered = this.parent.registerAgencyConfig(config);
    this.applyAutoWiring(config, registered as unknown as DocBuilderLike);
    return registered;
  }

  private applyAutoWiring(config: AgencyConfig, parent: DocBuilderLike): void {
    if (config.statusPath) {
      parent.field(config.statusPath, 'pending');
    }

    const workflowPrefix = `agency${config.token}`;
    const requestPermissionWorkflow: StepsCustomizer = (steps) => {
      const workerAgencyPermissions: Record<string, unknown> = {};
      if (config.allowedTypes.length > 0) {
        workerAgencyPermissions.allowedDocumentTypes = config.allowedTypes.map(
          (type) => ({ type }),
        );
      }
      if (config.allowedOperations.length > 0) {
        workerAgencyPermissions.allowedOperations = [
          ...config.allowedOperations,
        ];
      }
      steps
        .myOs()
        .grantWorkerAgencyPermission(
          config.onBehalfOf,
          config.requestId,
          workerAgencyPermissions,
          config.targetSessionId,
        );
    };

    if (config.permissionTiming === 'onInit') {
      parent.onInit(
        `${workflowPrefix}RequestPermission`,
        requestPermissionWorkflow,
      );
    } else if (config.permissionTiming === 'onEvent') {
      parent.onEvent(
        `${workflowPrefix}RequestPermission`,
        config.permissionTriggerEventType ?? 'Conversation/Event',
        requestPermissionWorkflow,
      );
    } else if (config.permissionTiming === 'onDocChange') {
      parent.onDocChange(
        `${workflowPrefix}RequestPermission`,
        config.permissionTriggerDocPath ?? '/',
        requestPermissionWorkflow,
      );
    }

    parent.onMyOsResponse(
      `${workflowPrefix}Granted`,
      'MyOS/Worker Agency Permission Granted',
      config.requestId,
      (steps) => {
        if (config.statusPath) {
          steps.replaceValue('MarkAgencyGranted', config.statusPath, 'granted');
        }
      },
    );

    parent.onMyOsResponse(
      `${workflowPrefix}Rejected`,
      'MyOS/Worker Agency Permission Rejected',
      config.requestId,
      (steps) => {
        if (config.statusPath) {
          steps.replaceValue(
            'MarkAgencyRejected',
            config.statusPath,
            'rejected',
          );
        }
      },
    );

    parent.onMyOsResponse(
      `${workflowPrefix}Revoked`,
      'MyOS/Worker Agency Permission Revoked',
      config.requestId,
      (steps) => {
        if (config.statusPath) {
          steps.replaceValue('MarkAgencyRevoked', config.statusPath, 'revoked');
        }
      },
    );
  }
}
